src/controller/event/handler/internal/artifact.go

```diff
@@ -55,43 +99,136 @@ func (a *Handler) IsStateful() bool {
+	a.once.Do(func() {
+		if !config.PullTimeUpdateDisable(ctx) {
+			a.pullTimeStore = make(map[string]time.Time)
+			go a.asyncFlushPullTime(orm.Context())
+		}
+
+		if !config.PullCountUpdateDisable(ctx) {
+			a.pullCountStore = make(map[int64]uint64)
+			go a.asyncFlushPullCount(orm.Context())
+		}
+	})
+
 	if !config.PullTimeUpdateDisable(ctx) {
-		go func() { a.updatePullTime(ctx, event) }()
+		a.updatePullTimeInCache(ctx, event)
 	}
+
 	if !config.PullCountUpdateDisable(ctx) {
-		go func() { a.addPullCount(ctx, event) }()
+		a.addPullCountInCache(ctx, event)
 	}
+
 	return nil
 }
 
-func (a *Handler) updatePullTime(ctx context.Context, event *event.ArtifactEvent) {
-	var tagID int64
+func (a *Handler) updatePullTimeInCache(ctx context.Context, event *event.ArtifactEvent) {
+	var tagName string
 	if len(event.Tags) != 0 {
-		tags, err := tag.Ctl.List(ctx, &q.Query{
-			Keywords: map[string]interface{}{
-				"ArtifactID": event.Artifact.ID,
-				"Name":       event.Tags[0],
-			},
-		}, nil)
+		tagName = event.Tags[0]
+	}
+
+	key := fmt.Sprintf("%d:%s", event.Artifact.ID, tagName)
+
+	a.pullTimeLock.Lock()
+	defer a.pullTimeLock.Unlock()
+
+	a.pullTimeStore[key] = time.Now()
+}
+
+func (a *Handler) addPullCountInCache(ctx context.Context, event *event.ArtifactEvent) {
+	a.pullCountLock.Lock()
+	defer a.pullCountLock.Unlock()
+
+	a.pullCountStore[event.Artifact.RepositoryID] = a.pullCountStore[event.Artifact.RepositoryID] + 1
+}
+
+func (a *Handler) syncFlushPullTime(ctx context.Context, artifactID int64, tagName string, time time.Time) {
+	var tagID int64
+
+	if tagName != "" {
+		tags, err := tag.Ctl.List(ctx, q.New(
+			map[string]interface{}{
+				"ArtifactID": artifactID,
+				"Name":       tagName,
+			}), nil)
 		if err != nil {
-			log.Infof("failed to list tags when to update pull time, %v", err)
+			log.Warningf("failed to list tags when to update pull time, %v", err)
 		} else {
 			if len(tags) != 0 {
 				tagID = tags[0].ID
 			}
 		}
 	}
-	if err := artifact.Ctl.UpdatePullTime(ctx, event.Artifact.ID, tagID, time.Now()); err != nil {
-		log.Debugf("failed to update pull time form artifact %d, %v", event.Artifact.ID, err)
+
+	if err := artifact.Ctl.UpdatePullTime(ctx, artifactID, tagID, time); err != nil {
+		log.Warningf("failed to update pull time for artifact %d, %v", artifactID, err)
+	}
+}
+
+func (a *Handler) syncFlushPullCount(ctx context.Context, repositoryID int64, count uint64) {
+	if err := repository.Ctl.AddPullCount(ctx, repositoryID, count); err != nil {
+		log.Warningf("failed to add pull count repository %d, %v", repositoryID, err)
 	}
-	return
 }
 
-func (a *Handler) addPullCount(ctx context.Context, event *event.ArtifactEvent) {
-	if err := repository.Ctl.AddPullCount(ctx, event.Artifact.RepositoryID); err != nil {
-		log.Debugf("failed to add pull count repository %d, %v", event.Artifact.RepositoryID, err)
+func (a *Handler) asyncFlushPullTime(ctx context.Context) {
```

Comment:
generally, we should consider the listen os shutdown signal in the goroutine to ensure update?